### ISSUE TITLE
another case of visual side effects

### ISSUE THREAD

**Issue by pkra** [zorkow/speech-rule-engine#462]
Not-at-all minimal example: 

```tex
\begin{equation} \bigg |\Big (\frac {\partial }{\partial r}\Big )^{\beta }\Big \langle { \frac {\partial ^{\alpha }\hat{f}}{\partial \xi _d^{\alpha }}\psi },{ \frac {\partial ^{\alpha }\hat{g}}{\partial \xi _d^{\alpha }}\psi }\Big \rangle _{\dot{H}^{-\gamma }(\Sigma _r)}\bigg |_{r=0}\bigg | \lesssim \|f\|_{L_p}\|g\|_{L_p}. \tag{49} \end{equation}
```

On mathjax.org#demo, we get

Without enrichment:

![image](https://user-images.githubusercontent.com/426686/108100680-8852fa00-7086-11eb-9ace-d3fb3ec94fa8.png)

With enrichment:

![image](https://user-images.githubusercontent.com/426686/108100716-943ebc00-7086-11eb-88c6-5f3292226b31.png)

Note how the subscripts are pushed down due to enrichment, the second falling outside it vertical bar.

I suspect this is expected behavior (due to additional mrows) but I wanted to share it as an example how these effects may add up.

**Comment by zorkow** [comment]
Somewhat expected behaviour. But the example exposes a number of other issues:

- [ ] Bracketing algorithm is greedy from the left. The evaluation bar is therefore taken to be the neutral fence. You either get the left hand side correct or the right hand side. 
- [ ] Converting the left-over bar to punctuation breaks up the relation. The heuristic to convert to pre-/postfix operators (part of the bra-ket heuristic) should help.
- [ ] The metric bars should probably not be spoken as parallel.

**Comment by pkra** [comment]
With more testing, I've found more and more cases, some considerably worse than this one - bad enough that I'm starting to consider this a bug. Couldn't SRE+MathJax work out a way to group things for SRE's annotations without affecting layout like this?

**Comment by zorkow** [comment]
Question: Are these cases exclusively related to the bracketing algorithm, i.e., stem from introduction of new rows? Or are there any other cases?

**Comment by pkra** [comment]
> Question: Are these cases exclusively related to the bracketing algorithm, i.e., stem from introduction of new rows? Or are there any other cases?

I'd have to look very carefully at the enriched output but it's my impression that it's always due tot additional rows.

**Comment by dpvc** [comment]
I suppose one possibility would be something MathJax output jax could do:  if an mrow is an mrow for fences (which we already mark), MathJax could use the final delimiter's size rather than the size of the mrow itself when attaching scripts.  That way, if the size of the delimiter is fixed size (by `\Bigr` or something), that size will be used, and if it is allowed to stretch, then it should get the size of the content anyway.  That might produce better results while still getting the grouping that enrichment produces.

The size information might also be used to help make proper delimiter matches:  if the open delimiter has a `minsize` or `maxsize`, only match it with a close delimiter with the same size.  That wouldn't help in this case since the middle bar is the same size as the outer ones, but it might help in other cases, such as the following abomination:

```
\Bigl| x - 3|1-x| \Bigr|
```

<img width="167" alt="abs-val" src="https://user-images.githubusercontent.com/432782/109652447-cdb70300-7b2d-11eb-8d17-843993428651.png">

One could also use the `l` and `r` from `\Bigl` and `\Bigr`, which ends up adding `texClass` of `OPEN` and `CLOSE` (but it is on an `mrow` surrounding the operator, since it come from an internal TeXAtom node).  The `texClass` could be added on the `mo` itself, if that helps.  Again, that doesn't resolve the issue in this expression, since the author only used `\Big` and `\bigg` rather than the left- and right-hand versions of these.

**Comment by dpvc** [comment]
> it's my impression that it's always due to additional rows.

Is it always about placement of super- or subscripts, or is it other spacing as well?

**Comment by zorkow** [comment]
If the case only occurs for fences, then we can make use of the `data-semantic-fencepointer`, which points to the fence that was the one actually embellished. I've just tried to a solution by post-processing the expression to put the embellishment back, while keeping the semantics in place.
Naturally this breaks the walker at the moment, but it also leads to some other visual anomalies for complex expressions.

**Comment by zorkow** [comment]
As an alternative, we could take care of that on the MathJax site:
When computing the `vertical-align` it could use the pointer argument a reference rather than the actual one.

**Comment by zorkow** [comment]
Here's an example: `(\frac{a}{b})^2`

This is what we currently get

```html
<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <msup DS-type="superscript" DS-role="leftright" DS-fencepointer="4" DS-id="6" DS-children="7,5">
    <mrow xmlns="http://www.w3.org/1999/xhtml" DS-type="fenced" DS-role="leftright" DS-id="7" DS-children="3" DS-content="0,4" DS-parent="6">
      <mo xmlns="http://www.w3.org/1998/Math/MathML" stretchy="false" DS-type="fence" DS-role="open" DS-id="0" DS-parent="7">(</mo>
      <mfrac xmlns="http://www.w3.org/1998/Math/MathML" DS-type="fraction" DS-role="division" DS-id="3" DS-children="1,2" DS-parent="7">
        <mi DS-type="identifier" DS-role="latinletter" DS-font="italic" DS-annotation="clearspeak:simple" DS-id="1" DS-parent="3">a</mi>
        <mi DS-type="identifier" DS-role="latinletter" DS-font="italic" DS-annotation="clearspeak:simple" DS-id="2" DS-parent="3">b</mi>
      </mfrac>
      <mo xmlns="http://www.w3.org/1998/Math/MathML" stretchy="false" DS-type="fence" DS-role="close" DS-id="4" DS-parent="7">)</mo>
    </mrow>
    <mn DS-type="number" DS-role="integer" DS-font="normal" DS-annotation="clearspeak:simple" DS-id="5" DS-parent="6">2</mn>
  </msup>
</math>
```

This is what we get with postprocessing:


```html
<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <mrow xmlns="http://www.w3.org/1999/xhtml" DS-type="fenced" DS-role="leftright" DS-id="7" DS-children="3" DS-content="0,4" DS-parent="6">
    <mo xmlns="http://www.w3.org/1998/Math/MathML" stretchy="false" DS-type="fence" DS-role="open" DS-id="0" DS-parent="7">(</mo>
    <mfrac xmlns="http://www.w3.org/1998/Math/MathML" DS-type="fraction" DS-role="division" DS-id="3" DS-children="1,2" DS-parent="7">
      <mi DS-type="identifier" DS-role="latinletter" DS-font="italic" DS-annotation="clearspeak:simple" DS-id="1" DS-parent="3">a</mi>
      <mi DS-type="identifier" DS-role="latinletter" DS-font="italic" DS-annotation="clearspeak:simple" DS-id="2" DS-parent="3">b</mi>
    </mfrac>
    <msup xmlns="http://www.w3.org/1998/Math/MathML" DS-type="superscript" DS-role="leftright" DS-fencepointer="4" DS-id="6" DS-children="7,5">
      <mo stretchy="false" DS-type="fence" DS-role="close" DS-id="4" DS-parent="7">)</mo>
      <mn DS-type="number" DS-role="integer" DS-font="normal" DS-annotation="clearspeak:simple" DS-id="5" DS-parent="6">2</mn>
    </msup>
  </mrow>
</math>
```

That works well enough for the simply case, but not in the complex one, which leads to subscripts being too high.

**Comment by pkra** [comment]
Thanks for taking a closer look!

> Is it always about placement of super- or subscripts, or is it other spacing as well?

The cases I noticed were only super/subscripts.

**Comment by dpvc** [comment]
@zorkow, your approach has promise, but doesn't that give the wrong semantics?  It's the whole delimited expression that is squared, not just the parenthesis, so isn't your original the proper semantics?  I'd rather consider improving the rendering on MathJax's end rather than confusing the semantics.

> The cases I noticed were only super/subscripts.

If this is about vertical placement, then I think it should not be too hard to handle it in MathJax output.  I will do some experiments.

**Comment by zorkow** [comment]
@dpvc Have you seen the draft [PR 605](https://github.com/mathjax/MathJax-src/pull/605)?
I've also finished `msubsup` and `mmultiscripts` cases last night while procrastinating on preparing a talk for today.
I'll push those commits and we can discuss over on the MJ repo.

**Comment by zorkow** [comment]
> If this is about vertical placement, then I think it should not be too hard to handle it in MathJax output. I will do some experiments.

Exactly, that's what PR 605 does. Only the vertical alignment is changed no change in semantics necessary.

**Comment by dpvc** [comment]
> Exactly, that's what PR 605 does.

Sorry, hadn't looked at that yet.  Yes, that is the sort of thing I had in mind.  Thanks for working it out!

**Comment by zorkow** [comment]
> > Exactly, that's what PR 605 does.
> 
> Sorry, hadn't looked at that yet. Yes, that is the sort of thing I had in mind. Thanks for working it out!

There's a few points where I am sure you can improve on my solution:
* It would be nice if everything could be done in the common wrapper instead of the single renderers. But I could not find a way.
* I also failed to work out the right mixin type, hence a lot of casting to any.

**Comment by dpvc** [comment]
@zorkow:

> It would be nice if everything could be done in the common wrapper instead of the single renderers. But I could not find a way.

Right.  I'll look into that.  Some refactoring may be in order.

> I also failed to work out the right mixin type, hence a lot of casting to any.

Yes, I saw that.  The secret usually is that you need to add your methods to the interface above the mixin.

**Comment by zorkow** [comment]
@dpvc The last two commits for `msubsup` and `mmultiscripts` have somehow ruined the treatment of simple sub and superscripts. Can you ignore those.
